# ***TicTok Project - Differentiating Opinions vc Claims***
<br/>

**The Project Goal** In order to reduce a large backlog of user generated reports that are, which are currently reviewed and moderated manually, we aim to develop a machine learning predictive model to classify claims vs opinions made in videos submitted to platforms. This is expected to reduce backlog and prioritize these contents more efficiently. 

**The Dataset Source** Google Analytics Course - Coursera



## Stage 1 - Data Cleaning and Analysis
**The Purpose** This Jupyter Notebook will investigate and understand the provided data. 

**Part 1** - Understanding the situation and goal.

**Part 2** - Understanding the data.

**Part 3** - Understanding the variables.

### Rough notes
1. include a summary of the column Data types, data value nonnull counts, relevant and irrelevant columns, along with anything else code related you think is worth sharing/showing in the notebook? You’ll need to select a couple of variables to focus on. Include their minimum and maximum values.

In [1]:
# importing necessary libraries
import pandas as pd 
import numpy as np 

In [2]:
# Reading the csv file
data_main = pd.read_csv("tiktok_dataset.csv")

### Part 2 - Understanding the data and inspecting the dataset

In [3]:
data_main.head(10)

# claim_status    video_id  video_duration_sec  \
0   1        claim  7017666017                  59   
1   2        claim  4014381136                  32   
2   3        claim  9859838091                  31   
3   4        claim  1866847991                  25   
4   5        claim  7105231098                  19   
5   6        claim  8972200955                  35   
6   7        claim  4958886992                  16   
7   8        claim  2270982263                  41   
8   9        claim  5235769692                  50   
9  10        claim  4660861094                  45   

                            video_transcription_text verified_status  \
0  someone shared with me that drone deliveries a...    not verified   
1  someone shared with me that there are more mic...    not verified   
2  someone shared with me that american industria...    not verified   
3  someone shared with me that the metro of st. p...    not verified   
4  someone shared with me that the number of busi...    not verified   
5  someone shared with me that gross domestic pro...    not verified   
6  someone shared with me that elvis presley has ...    not verified   
7  someone shared with me that the best selling s...    not verified   
8  someone shared with me that about half of the ...    not verified   
9  someone shared with me that it would take a 50...        verified   

  author_ban_status  video_view_count  video_like_count  video_share_count  \
0      under review          343296.0           19425.0              241.0   
1            active          140877.0           77355.0            19034.0   
2            active          902185.0           97690.0             2858.0   
3            active          437506.0          239954.0            34812.0   
4            active           56167.0           34987.0             4110.0   
5      under review          336647.0          175546.0            62303.0   
6            active          750345.0          486192.0           193911.0   
7            active          547532.0            1072.0               50.0   
8            active           24819.0           10160.0             1050.0   
9            active          931587.0          171051.0            67739.0   

   video_download_count  video_comment_count  
0                   1.0                  0.0  
1                1161.0                684.0  
2                 833.0                329.0  
3                1234.0                584.0  
4                 547.0                152.0  
5                4293.0               1857.0  
6                8616.0               5446.0  
7                  22.0                 11.0  
8                  53.0                 27.0  
9                4104.0               2540.0

In [4]:
print(f"Shape of dataset is {data_main.shape}")

Shape of dataset is (19382, 12)


This dataset contains 19382 rows and 12 columns. The column names and their description are as follows :
1. "#" - A number assigned to each video by TicTok.
2. "claim_status" - If a video is identified as a claim or an opinion.
3. "video_id" - Random identifying number assigned to a video when published on TicTok.
4. "video_duration_sec" - The length of the video in seconds.
5. "video_transcription_text" - Transcribed texts of words spoken in video
6. "verified_status" -  The status of the maker of video, if they are a verified user on TicTok or not. 
7. "author_ban_status" - Indicating the permissions of the account of the author who published the video. 
8. "video_view_count" - Number of views obtained on the video.
9. "video_like_count" - Number of likes obtained on the video.
10. "video_share_count" - Number of times the video was shared.
11. "video_download_count" - Number of times the video was downloaded.
12. "video_comment_count" - Number of comments obtained on the video.

In [5]:
data_main.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19382 entries, 0 to 19381
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   #                         19382 non-null  int64  
 1   claim_status              19084 non-null  object 
 2   video_id                  19382 non-null  int64  
 3   video_duration_sec        19382 non-null  int64  
 4   video_transcription_text  19084 non-null  object 
 5   verified_status           19382 non-null  object 
 6   author_ban_status         19382 non-null  object 
 7   video_view_count          19084 non-null  float64
 8   video_like_count          19084 non-null  float64
 9   video_share_count         19084 non-null  float64
 10  video_download_count      19084 non-null  float64
 11  video_comment_count       19084 non-null  float64
dtypes: float64(5), int64(3), object(4)
memory usage: 1.8+ MB


In [6]:
data_main.isna().sum()

#                             0
claim_status                298
video_id                      0
video_duration_sec            0
video_transcription_text    298
verified_status               0
author_ban_status             0
video_view_count            298
video_like_count            298
video_share_count           298
video_download_count        298
video_comment_count         298
dtype: int64

This dataset contains nulls in columns : claim_status, video_transcription_text, video_view_count, video_like_count,video_share_count, video_download_count and video_comment_count.
<br/>

The dataset columns are a combination of numeric and categorical data types. As an observation, all the columns that have null values happen to have the same number of null values. 

**Question** What kind of data would the rows with null values indicate? Are the videos duplicated? Are the authors duplicated? Are the videos deleted?

In [7]:
data_main.describe()

#      video_id  video_duration_sec  video_view_count  \
count  19382.000000  1.938200e+04        19382.000000      19084.000000   
mean    9691.500000  5.627454e+09           32.421732     254708.558688   
std     5595.245794  2.536440e+09           16.229967     322893.280814   
min        1.000000  1.234959e+09            5.000000         20.000000   
25%     4846.250000  3.430417e+09           18.000000       4942.500000   
50%     9691.500000  5.618664e+09           32.000000       9954.500000   
75%    14536.750000  7.843960e+09           47.000000     504327.000000   
max    19382.000000  9.999873e+09           60.000000     999817.000000   

       video_like_count  video_share_count  video_download_count  \
count      19084.000000       19084.000000          19084.000000   
mean       84304.636030       16735.248323           1049.429627   
std       133420.546814       32036.174350           2004.299894   
min            0.000000           0.000000              0.000000   
25%          810.750000         115.000000              7.000000   
50%         3403.500000         717.000000             46.000000   
75%       125020.000000       18222.000000           1156.250000   
max       657830.000000      256130.000000          14994.000000   

       video_comment_count  
count         19084.000000  
mean            349.312146  
std             799.638865  
min               0.000000  
25%               1.000000  
50%               9.000000  
75%             292.000000  
max            9599.000000

1. Columns # and video_id are numeric but do not have values are describe numeric entries. Therefore, their statistics descriptions shall be meaningless. 

2. Videos on TikToc for this dataset range from 5 seconds long to 60 seconds long with average length being 32.4 seconds. This column appears to be normally distributed. 

3. Minimum a video is viewed is 20 times to a maximum of 999,817 times with an average of 254,708.56 and median lying on 9,954.5 (50th percentile). This is a rightly skewed data with maximum of its values lying towards left of mean. This is because the average is relatively high compared to the minimum, suggesting that there are some videos with extremely high view counts pulling the average up, while many videos might have lower views. Also, the large difference between the 75th percentile (504,327 views) and the maximum (999,817 views) further confirms this right skew. Most videos have relatively low views, but a few very popular ones have very high views, pulling the mean upwards.

4. On observing data from other columns, we can conclude that the data is rightly skewed because of similar behavior of other columns.

5. This is a clear indication of outliers being present in the data. 

### Claim Statuses - Options or Claims

In [8]:
data_main['claim_status'].value_counts()

claim_status
claim      9608
opinion    9476
Name: count, dtype: int64

In [9]:
# Claim status : Opinion or Claim
data_main.groupby('claim_status').count()

#  video_id  video_duration_sec  video_transcription_text  \
claim_status                                                                 
claim         9608      9608                9608                      9608   
opinion       9476      9476                9476                      9476   

              verified_status  author_ban_status  video_view_count  \
claim_status                                                         
claim                    9608               9608              9608   
opinion                  9476               9476              9476   

              video_like_count  video_share_count  video_download_count  \
claim_status                                                              
claim                     9608               9608                  9608   
opinion                   9476               9476                  9476   

              video_comment_count  
claim_status                       
claim                        9608  
opinion                      9476

There are 9608 videos that are labeled as claims and 9476 that are labeled as opinion. This implies, there are 298 rows with no labels, as stated above and there are more videos labeled as claims than as opinions.

In [10]:
mask_claim = data_main[data_main['claim_status'] == 'claim']
mask_opinion = data_main[data_main['claim_status'] == 'opinion']

In [11]:
print(f"The average view count of videos with claim as status is {mask_claim['video_view_count'].mean()}.")
print(f"The average view count of videos with opinion as status is {mask_opinion['video_view_count'].mean()}.")

The average view count of videos with claim as status is 501029.4527477102.
The average view count of videos with opinion as status is 4956.43224989447.


In [12]:
print(f"The median view of videos with claim as status is {mask_claim['video_view_count'].median()}.")
print(f"The median view of videos with opinion as status is {mask_opinion['video_view_count'].median()}.")

The median view of videos with claim as status is 501555.0.
The median view of videos with opinion as status is 4953.0.


The average views on videos with status as claim is 501,029.45 and that on videos with status as opinion is 4,956.43 which is 496073.02049781574 more views on videos with status claim, even though there are only 132 claim status videos more than opinion status videos. 

This suggests that videos with "claim" status tend to attract significantly more attention or engagement compared to those labeled as "opinion." However, it's worth exploring why "claim" videos might get more views—factors such as content type, topic relevance, or viewer interest could contribute to this disparity.

The median view on videos with status as claim is 501,555 views and that on videos with status as opinion is 4,953 views. These values are very close to their respective mean values.

Thus, individually, both the distributions are likely to be symmetric.



### Statuses of Authors - Active, Banned and Under review

In [13]:
# Ban status of Author
data_main.groupby('author_ban_status').count()[['#']]

#
author_ban_status       
active             15663
banned              1639
under review        2080

Out of 19382 rows with statuses about authors, 15,663 are active accounts, 1,639 are banned and 2,080 are under review.

In [14]:
mask_active = data_main[data_main['author_ban_status'] == 'active']
mask_ban = data_main[data_main['author_ban_status'] == 'banned']
mask_review = data_main[data_main['author_ban_status'] == 'under review']

In [15]:
print(f"The average view count of videos with authors having active account is {mask_active['video_view_count'].mean()}.")
print(f"The average view count of videos with authors having banned account is {mask_ban['video_view_count'].mean()}.")
print(f"The average view count of videos with authors having account that are under review is {mask_review['video_view_count'].mean()}.")

The average view count of videos with authors having active account is 215927.03952415002.
The average view count of videos with authors having banned account is 445845.4391437309.
The average view count of videos with authors having account that are under review is 392204.8363988383.


In [16]:
print(f"The median view of videos with authors having active account is {mask_active['video_view_count'].median()}.")
print(f"The median view of videos with authors having banned account is {mask_ban['video_view_count'].median()}.")
print(f"The median view of videos with authors having account that are under review is {mask_review['video_view_count'].median()}.")

The median view of videos with authors having active account is 8616.0.
The median view of videos with authors having banned account is 448201.0.
The median view of videos with authors having account that are under review is 365245.5.


The average engagement values of active accounts, banned accounts and accounts under review are 215,927.03, 445,845.44 and 392,204.84 respectively. Similarly, their median viewership values are 8,616, 448,201 and 365,245.5. These values suggest that the banned author accounts receive the maximum viewership and engagement with accounts that are under review as second and authors with active account as last. 

The large difference between the mean engagement (215,927.03) and median viewership (8,616) for active accounts indicates possible skewness. The mean being significantly higher than the median suggests that a few outliers with very high engagement are inflating the average. In contrast, the data for banned accounts and accounts under review seems more consistent, with their medians relatively close to their mean values.

In [17]:
data_main.groupby(['claim_status', 'author_ban_status'])[['video_view_count']].mean(numeric_only=True)

video_view_count
claim_status author_ban_status                  
claim        active                499221.733171
             banned                505907.917304
             under review          504054.640674
opinion      active                  4958.120563
             banned                  4876.530612
             under review            4958.105832

In [18]:
data_main.groupby(['claim_status', 'author_ban_status'])[['video_id']].count()

video_id
claim_status author_ban_status          
claim        active                 6566
             banned                 1439
             under review           1603
opinion      active                 8817
             banned                  196
             under review            463

Out of 1639 total banned videos, 1439 banned videos are given status as claim and only 196 videos have status as opinion. This means videos that state to be claims are strictly policed than opinion videos. 

In [19]:
data_main.groupby('author_ban_status')[['video_share_count']].median()

video_share_count
author_ban_status                   
active                         437.0
banned                       14468.0
under review                  9444.0

In [20]:
data_main.groupby('author_ban_status')[['video_share_count']].mean()

video_share_count
author_ban_status                   
active                  14111.466164
banned                  29998.942508
under review            25774.696999

The median of videos shared of banned account is 14,468 while that of active is only 437. That is 14,031 more than active.

In [21]:
data_main.groupby('author_ban_status').agg({'video_view_count': ['count','mean','median'],
                                           'video_like_count' : ['count','mean','median'],
                                           'video_share_count' : ['count','mean','median']})

video_view_count                          video_like_count  \
                             count           mean    median            count   
author_ban_status                                                              
active                       15383  215927.039524    8616.0            15383   
banned                        1635  445845.439144  448201.0             1635   
under review                  2066  392204.836399  365245.5             2066   

                                           video_share_count                \
                            mean    median             count          mean   
author_ban_status                                                            
active              71036.533836    2222.0             15383  14111.466164   
banned             153017.236697  105573.0              1635  29998.942508   
under review       128718.050339   71204.5              2066  25774.696999   

                            
                    median  
author_ban_status           
active               437.0  
banned             14468.0  
under review        9444.0

In [22]:
data_main.groupby(['claim_status','author_ban_status']).agg({'video_view_count': ['count','mean','median'],
                                           'video_like_count' : ['count','mean','median'],
                                           'video_share_count' : ['count','mean','median']})

video_view_count                           \
                                          count           mean    median   
claim_status author_ban_status                                             
claim        active                        6566  499221.733171  499500.5   
             banned                        1439  505907.917304  512572.0   
             under review                  1603  504054.640674  500774.0   
opinion      active                        8817    4958.120563    4958.0   
             banned                         196    4876.530612    5083.5   
             under review                   463    4958.105832    4884.0   

                               video_like_count                           \
                                          count           mean    median   
claim_status author_ban_status                                             
claim        active                        6566  164960.302924  121943.0   
             banned                        1439  173719.102849  132044.0   
             under review                  1603  165566.954460  125882.0   
opinion      active                        8817    1091.714982     820.0   
             banned                         196    1027.515306     799.5   
             under review                   463    1139.663067     876.0   

                               video_share_count                         
                                           count          mean   median  
claim_status author_ban_status                                           
claim        active                         6566  32769.101889  17774.5  
             banned                         1439  34056.580959  19018.0  
             under review                   1603  33155.623206  18084.0  
opinion      active                         8817    217.166950    121.0  
             banned                          196    208.423469    108.5  
             under review                    463    220.431965    124.0

The number of views, likes and shares on videos for banned authors is less than that of active authors but their average values and median values are much more than the active authors.

We observe that videos from banned authors and those under review generally receive significantly more views, likes, and shares compared to videos from non-banned authors. Nonetheless, *when a video is viewed*, its engagement rate is more influenced by its claim status than by the author’s ban status.

Additionally, claim videos not only attract a higher view rate compared to opinion videos, but they also receive more likes on average, indicating they are better received overall. Moreover, claim videos see more engagement through comments and shares than opinion videos.

For claim videos, banned authors have slightly higher likes per view and shares per view rates compared to active authors and those under review. However, for opinion videos, active authors and those under review generally achieve higher engagement rates than banned authors across all categories.

In [23]:
data_new = data_main
data_new['likes_per_view'] = data_main['video_like_count'] / data_main['video_view_count']
data_new['comments_per_view'] = data_main['video_comment_count'] / data_main['video_view_count']
data_new['shares_per_view'] = data_main['video_share_count'] / data_main['video_view_count']
data_new.head(10)

# claim_status    video_id  video_duration_sec  \
0   1        claim  7017666017                  59   
1   2        claim  4014381136                  32   
2   3        claim  9859838091                  31   
3   4        claim  1866847991                  25   
4   5        claim  7105231098                  19   
5   6        claim  8972200955                  35   
6   7        claim  4958886992                  16   
7   8        claim  2270982263                  41   
8   9        claim  5235769692                  50   
9  10        claim  4660861094                  45   

                            video_transcription_text verified_status  \
0  someone shared with me that drone deliveries a...    not verified   
1  someone shared with me that there are more mic...    not verified   
2  someone shared with me that american industria...    not verified   
3  someone shared with me that the metro of st. p...    not verified   
4  someone shared with me that the number of busi...    not verified   
5  someone shared with me that gross domestic pro...    not verified   
6  someone shared with me that elvis presley has ...    not verified   
7  someone shared with me that the best selling s...    not verified   
8  someone shared with me that about half of the ...    not verified   
9  someone shared with me that it would take a 50...        verified   

  author_ban_status  video_view_count  video_like_count  video_share_count  \
0      under review          343296.0           19425.0              241.0   
1            active          140877.0           77355.0            19034.0   
2            active          902185.0           97690.0             2858.0   
3            active          437506.0          239954.0            34812.0   
4            active           56167.0           34987.0             4110.0   
5      under review          336647.0          175546.0            62303.0   
6            active          750345.0          486192.0           193911.0   
7            active          547532.0            1072.0               50.0   
8            active           24819.0           10160.0             1050.0   
9            active          931587.0          171051.0            67739.0   

   video_download_count  video_comment_count  likes_per_view  \
0                   1.0                  0.0        0.056584   
1                1161.0                684.0        0.549096   
2                 833.0                329.0        0.108282   
3                1234.0                584.0        0.548459   
4                 547.0                152.0        0.622910   
5                4293.0               1857.0        0.521454   
6                8616.0               5446.0        0.647958   
7                  22.0                 11.0        0.001958   
8                  53.0                 27.0        0.409364   
9                4104.0               2540.0        0.183612   

   comments_per_view  shares_per_view  
0           0.000000         0.000702  
1           0.004855         0.135111  
2           0.000365         0.003168  
3           0.001335         0.079569  
4           0.002706         0.073175  
5           0.005516         0.185069  
6           0.007258         0.258429  
7           0.000020         0.000091  
8           0.001088         0.042306  
9           0.002727         0.072714

In [24]:
data_agg = data_new.groupby(['claim_status','author_ban_status']).agg({'likes_per_view' : ['count','mean','median'],
                                                                       'comments_per_view' : ['count','mean','median'],
                                                                       'shares_per_view' : ['count','mean','median']})

In [25]:
data_agg

likes_per_view                      \
                                        count      mean    median   
claim_status author_ban_status                                      
claim        active                      6566  0.329542  0.326538   
             banned                      1439  0.345071  0.358909   
             under review                1603  0.327997  0.320867   
opinion      active                      8817  0.219744  0.218330   
             banned                       196  0.206868  0.198483   
             under review                 463  0.226394  0.228051   

                               comments_per_view                      \
                                           count      mean    median   
claim_status author_ban_status                                         
claim        active                         6566  0.001393  0.000776   
             banned                         1439  0.001377  0.000746   
             under review                   1603  0.001367  0.000789   
opinion      active                         8817  0.000517  0.000252   
             banned                          196  0.000434  0.000193   
             under review                    463  0.000536  0.000293   

                               shares_per_view                      
                                         count      mean    median  
claim_status author_ban_status                                      
claim        active                       6566  0.065456  0.049279  
             banned                       1439  0.067893  0.051606  
             under review                 1603  0.065733  0.049967  
opinion      active                       8817  0.043729  0.032405  
             banned                        196  0.040531  0.030728  
             under review                  463  0.044472  0.035027

* Out of 19382 rows, 9608 were claims, i.r. 49.57% and 9476 were opinions i.r. 48.89% and 298 were rows with no claims i.e 1.53%.

* Engagement level is strongly co-related with claim status. It is also co-related to author's ban status but slightly less.